### PR TITLE
Manage client snippet on server side

### PR DESCRIPTION
**Pull request by angelozerr**
Manage client snippet on server side

Fixes https://github.com/redhat-developer/vscode-quarkus/issues/119

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR manages Quarkus snippets on server side. Snippets are the same than vscode snippet except that it adds a `context` to show the  snippet according the project dependency. Ex : `qds` appears only if there is quarkus-agroal in the classpath:

```json
"context": {
	"extension": "quarkus-agroal"
}
```

For the moment it's a draft PR since I must comment all codes and write tests. You can play with this PR and you should see `qds`  twice (from vscode snippet and from MicroProfile LS) and compare the 2 behaviors. It should be the same except `qds` (from MicroProfile LS) should appear only if quarkus-agroal is on the classpath.

When this PR will be merged, we could methe the PR from vscode-quarkus https://github.com/redhat-developer/vscode-quarkus/pull/228 which removes those snippets on client side.

**Comment by angelozerr** [comment]
Here a demo, you can see that there one `qds` (client side) when there is not dependency with `quarkus-agroal`, when you add it, you should see two `qds` snippets (client and server side):

QDSSnippetDemo

To test this PR:

 * check that qds (server) appears only when project has `quarkus-agroal` dependency
 * check open qds completion on server side  has the same behavior than qds completion on client side (check qds are provided in same time if you type only, q, and qd, and other tests)
 * check apply completion is the same behavior for cliengt and server snippet.

Once this PR will be merged, we will able to to remove client snippets.

**Comment by angelozerr** [comment]
This PR includes too snippets for java files mp rest and mp openapi snippets written by @rzgry https://github.com/redhat-developer/quarkus-ls/issues/246#issuecomment-599714894

OperationSnippetDemo

@rzgry do you want I remove your snippets from this PR? Or is it OK to keep it and create a new PR to update them?

Today it should work like vscode snippet client, but I need to manage filter :

 * filter according dependency (show open api snippet only if project have mp pen api dependency)
 * filter according to the completion offset (ex: show operation snippet only if completion offset is before a class,method declaration)

**Comment by rzgry** [comment]
@angelozerr its fine for you to keep the snippets in this PR. I can open other PRs to update the existing ones / add snippets for other microprofile specifications

**Comment by angelozerr** [comment]
> @angelozerr its fine for you to keep the snippets in this PR. 

Nice, I will write some tests around those snippets. I think this PR will not manage filter, I would like to do that in an another PR. When we did that we will update snippets json to manage context.

**Comment by xorye** [comment]
The application.prooperties snippets work well on my machine.

For the Java snippets, I realized that the server side snippets do not adhere to the current indentation.

I noticed this when invoking a snippet inside of a class:
gif of indentation issue

whereas the vs code snippet follows the current indentation
gif of vs code snippet

**Comment by xorye** [comment]
Another comment about the Java snippets:
I'm not sure if this is a limitation with VS Code, but sometimes I have a difficult time invoking snippets like `@Metric`.

If VS Code provides completion after the `@` is typed, as I type `Metric`, the snippet does not appear.
snippet gif
To have the snippet appear, I have to press Escape to cancel the completion and invoke completion again.

**Comment by angelozerr** [comment]
> I realized that the server side snippets do not adhere to the current indentation.

It should be fixed.

> f VS Code provides completion after the @ is typed, as I type Metric, the snippet does not appear.

It should be fixed.

The PR is now ready, I write tests too.

The load of snippets is done by Java SPI (see https://github.com/redhat-developer/quarkus-ls/pull/251/files#diff-56d60c257da46fc9af0a48f34ca4798eR67) and for the moment Quarkus snippets are hosted into MicroProfile LS too. We need to create a new project `com.redhat.quarkus.ls` which provides by SPI those snippets, but I think we could do that when MicroProfile LS will move to Eclipse.org. @fbricon are you OK with that?

Today the snippet of application.properties are shown according to the dependency of the project. For java snippets for the momen tit is not done. I would like to manage that in an another PR.

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/JavaTextDocumentService.java [hunk 1]
```diff
@@ -30,16 +37,23 @@
...
 import com.redhat.microprofile.commons.DocumentFormat;
 import com.redhat.microprofile.commons.MicroProfileJavaCodeActionParams;
 import com.redhat.microprofile.commons.MicroProfileJavaCodeLensParams;
 import com.redhat.microprofile.commons.MicroProfileJavaDiagnosticsParams;
 import com.redhat.microprofile.commons.MicroProfileJavaHoverParams;
+import com.redhat.microprofile.ls.commons.BadLocationException;
+import com.redhat.microprofile.ls.commons.TextDocument;
+import com.redhat.microprofile.ls.commons.TextDocuments;
 import com.redhat.microprofile.ls.commons.client.CommandKind;
+import com.redhat.microprofile.ls.commons.snippets.SnippetRegistry;
```
Comment (xorye) [inline microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/JavaTextDocumentService.java]: Please remove unused import

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+/**
+ * GSON deserializer to build Snippt from vscode JSON snippet.
```
Comment (xorye) [inline microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetDeserializer.java]: ... build Snippet from ...

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetRegistry.java [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+			throws JsonIOException, JsonSyntaxException {
+		GsonBuilder builder = new GsonBuilder();
+		builder.registerTypeAdapter(Snippet.class, new SnippetDeserializer(contextDeserializer));
+		return builder.create().fromJson(reader, Snippet.class);
+	}
+
+	/**
+	 * Returns the all snippets.
+	 * 
+	 * @return the all snippets.
```
Comment (xorye) [inline microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetRegistry.java]: /**
 * Returns all snippets.
 * 
 * @return all snippets.

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/SnippetContextProperties.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+*******************************************************************************/
+package com.redhat.microprofile.snippets;
+
+import java.util.Set;
+
+import com.redhat.microprofile.ls.commons.snippets.ISnippetContext;
+
+/**
+ * A snippet context properties which matches extension. You can write a vscode
+ * snippet and declares a context like this:
```
Comment (xorye) [inline microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/SnippetContextProperties.java]: ... snippet that declares ...

**Comment by xorye** [comment]
The new fixes work great on my machine.

Please fix this failed test:
```
Running com.redhat.microprofile.services.ApplicationPropertiesCompletionTest
Tests run: 17, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.8 sec <<< FAILURE!
completionOnProfile(com.redhat.microprofile.services.ApplicationPropertiesCompletionTest)  Time elapsed: 0.007 sec  <<< FAILURE!
java.lang.AssertionError: expected:<3> but was:<5>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at com.redhat.microprofile.services.MicroProfileAssert.assertCompletions(MicroProfileAssert.java:176)
        at com.redhat.microprofile.services.MicroProfileAssert.testCompletionFor(MicroProfileAssert.java:161)
        at com.redhat.microprofile.services.MicroProfileAssert.testCompletionFor(MicroProfileAssert.java:122)
        at com.redhat.microprofile.services.ApplicationPropertiesCompletionTest.completionOnProfile(ApplicationPropertiesCompletionTest.java:138)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
```

**Comment by angelozerr** [comment]
> Please fix this failed test:

fixed.

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/SnippetContextProperties.java (hunk 4 above)
Comment (angelozerr) [reply microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/SnippetContextProperties.java]: fixed

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetRegistry.java (hunk 3 above)
Comment (angelozerr) [reply microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetRegistry.java]: fixed

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetDeserializer.java (hunk 2 above)
Comment (angelozerr) [reply microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/commons/snippets/SnippetDeserializer.java]: fixed

File: microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/JavaTextDocumentService.java (hunk 1 above)
Comment (angelozerr) [reply microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/JavaTextDocumentService.java]: fixed

File: microprofile.ls/README.md [hunk 5]
```diff
@@ -1,17 +1,20 @@
...
-
-The Quarkus language works alongside the
-[Quarkus jdt.ls extension](https://github.com/redhat-developer/quarkus-ls/tree/master/quarkus.jdt)
-which is also located in this repository. The Quarkus jdt.ls extension is able to index
-a Quarkus project for metadata (config properties, documentation, sources, etc.) to provide the 
-Quarkus language server the information required for the language features.
-The relationship between the Quarkus language server and the Quarkus jdt.ls extension is explained 
+language features for: 
+
+ * properties file: microprofile-config.properties, application.properties, application.yaml file from a
```
Comment (xorye) [inline microprofile.ls/README.md]: * Properties files: microprofile-config.properties, application.properties, application.yaml files from a
[MicroProfile](https://microprofile.io/) or [Quarkus](https://quarkus.io/) project.

File: microprofile.ls/README.md [hunk 6]
```diff
@@ -1,17 +1,20 @@
...
+[MicroProfile](https://microprofile.io/), [Quarkus](https://quarkus.io/) project.
+ * java files (URL CodeLens for JAX-RS, Java MicroProfile snippets, etc).
```
Comment (xorye) [inline microprofile.ls/README.md]: * Java files (URL CodeLens for JAX-RS, Java MicroProfile snippets, etc.).

File: microprofile.ls/README.md [hunk 7]
```diff
@@ -1,17 +1,20 @@
...
+
+The MicroProfile language works alongside the
```
Comment (xorye) [inline microprofile.ls/README.md]: The MicroProfile language server works alongside the

File: microprofile.ls/README.md [hunk 8]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
- * Visual Studio Code with [vscode-quarkus](https://github.com/redhat-developer/vscode-quarkus)
\ No newline at end of file
+ * Visual Studio Code with [vscode-quarkus](https://github.com/redhat-developer/vscode-quarkus)
+ 
+Code Snippets
+-------
+
+Java and properties completion snippets are managed by the MicroProfile LS (snippets on server side) by using Java SPI.
+
+## Describes snippets in JSON
```
Comment (xorye) [inline microprofile.ls/README.md]: ## Describing snippets in JSON

File: microprofile.ls/README.md [hunk 9]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+
+Snippets are described in a JSON files by following the [vscode snippet format](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_create-your-own-snippets).
```
Comment (xorye) [inline microprofile.ls/README.md]: Snippets are described in JSON files using the [vscode snippet format]

File: microprofile.ls/README.md [hunk 10]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+
+A context can be described to show the snippet according to a condition:
```
Comment (xorye) [inline microprofile.ls/README.md]: The `"context"` key can be provided to specify the condition in which the snippet should appear:

File: microprofile.ls/README.md [hunk 11]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+
+ * for properties file: the context/extension can be declared in the snippet to show the snippet only if the extension belongs to the project.
```
Comment (xorye) [inline microprofile.ls/README.md]: * for properties files: ...

File: microprofile.ls/README.md [hunk 12]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+		"body": [
+			...
+		],
+		"context": {
+			"extension": "quarkus-agroal"
+		}
+	}
+```
+
+means that the snippet is shown only if the project have the `quarkus-agroal` dependency.
```
Comment (xorye) [inline microprofile.ls/README.md]: ... if the project has the ...

File: microprofile.ls/README.md [hunk 13]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+    
+ * for java files (not available for the moment)
```
Comment (xorye) [inline microprofile.ls/README.md]: for Java files (not available for the moment)

File: microprofile.ls/README.md [hunk 14]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+ 
+## Add new internal snippets
```
Comment (xorye) [inline microprofile.ls/README.md]: ## Adding new internal snippets

File: microprofile.ls/README.md [hunk 15]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+
+To register a snippet you need to add it in :
+
+ * [MicroProfileJavaSnippetRegistryLoader](https://github.com/redhat-developer/quarkus-ls/blob/master/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/MicroProfileJavaSnippetRegistryLoader.java) if the new snippets is for java files.
```
Comment (xorye) [inline microprofile.ls/README.md]: ... if the new snippet is for Java files.

File: microprofile.ls/README.md [hunk 16]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+means that the snippet is shown only if the project have the `quarkus-agroal` dependency.
+    
+ * for java files (not available for the moment)
+ 
+## Add new internal snippets
+
+To register a snippet you need to add it in :
+
+ * [MicroProfileJavaSnippetRegistryLoader](https://github.com/redhat-developer/quarkus-ls/blob/master/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/MicroProfileJavaSnippetRegistryLoader.java) if the new snippets is for java files. 
+ * [MicroProfilePropertiesSnippetRegistryLoader](https://github.com/redhat-developer/quarkus-ls/blob/master/microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/snippets/MicroProfilePropertiesSnippetRegistryLoader.java) if the new snippets is for properties files.
```
Comment (xorye) [inline microprofile.ls/README.md]: ... if the new snippet is for properties files.

File: microprofile.ls/README.md [hunk 17]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+
+## Add new external snippets
```
Comment (xorye) [inline microprofile.ls/README.md]: ## Adding new external snippets

File: microprofile.ls/README.md [hunk 18]
```diff
@@ -43,8 +46,52 @@ Building the Language Server
...
+
+To add external snippets (like Quarkus) you need to create an implementation of ISnippetRegistryLoader and register it with Java SPI. See for Quarkus:
```
Comment (xorye) [inline microprofile.ls/README.md]: To add external snippets (like Quarkus snippets) an implementation of `ISnippetRegistryLoader` and register it with Java SPI must be created. See for Quarkus snippets:

File: microprofile.ls/README.md [hunk 19]
```diff
@@ -43,8 +45,52 @@ Building the Language Server
...
+	}
+```
+
+means that the snippet is shown only if the project has the `quarkus-agroal` dependency.
+    
+ * for `Java files` (not available for the moment)
+ 
+## Adding new internal snippets
+
+To register a snippet you need to add it in :
```
Comment (xorye) [inline microprofile.ls/README.md]: To register a snippet, it must be added in:

Comment (angelozerr) [reply microprofile.ls/README.md]: fixed

**Pull request merged**